packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp

```diff
@@ -3504,6 +3084,282 @@ merge_matrices(CrsMatrixStruct<Scalar, LocalOrdinal, GlobalOrdinal, Node>& Aview
+    {
+      if(finalPass && present.test(i))
+      {
+        gidList(lcount) = minGID + i;
+      }
+      if(present.test(i))
+      {
+        lcount++;
+      }
+    }
+
+    GO minGID;
+    GView gidList;
+    const_bitset_t present;
+  };
+
+  template<typename GO, typename GView, typename LView, typename LocalMapType>
+  struct ConvertGlobalToLocal
+  {
+    ConvertGlobalToLocal(const GView& gids_, LView& lids_, LocalMapType& localMap_)
+      : gids(gids_), lids(lids_), localMap(localMap_)
+    {}
+
+    KOKKOS_INLINE_FUNCTION void operator()(const GO i) const
+    {
+      lids(i) = localMap.getLocalElement(gids(i));
+    }
+
+    GView gids;
+    LView lids;
+    LocalMapType localMap;
+  };
+} //end ColMapFunctors
+
+//Build the minimal (sorted) column map for the given set of global columns
+//Then convert gids and store them in lids (gids is not modified)
+template<typename Scalar, typename LocalOrdinal, typename GlobalOrdinal, typename Node>
+Teuchos::RCP<Map<LocalOrdinal, GlobalOrdinal, Node> > AddKernels<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
+makeColMapAndConvertGids(GlobalOrdinal indexBase, GlobalOrdinal minCol, GlobalOrdinal maxCol,
+                   const typename MMdetails::AddKernels<Scalar, LocalOrdinal, GlobalOrdinal, Node>::global_col_inds_array& gids,
+                   typename MMdetails::AddKernels<Scalar, LocalOrdinal, GlobalOrdinal, Node>::col_inds_array& lids,
+                   const Teuchos::RCP<const Teuchos::Comm<int>>& comm)
+{
+  using Teuchos::RCP;
+  using Teuchos::rcp;
+  int myRank = comm->getRank();
+  std::cout << "Rank " << myRank << ": column map for some subset of global cols: " << minCol << " <= col <= " << maxCol << '\n';
+  typedef typename MMdetails::AddKernels<Scalar, LocalOrdinal, GlobalOrdinal, Node>::global_col_inds_array::non_const_type GView;
+  typedef typename MMdetails::AddKernels<Scalar, LocalOrdinal, GlobalOrdinal, Node>::col_inds_array::non_const_type LView;
+  typedef typename Map<LocalOrdinal, GlobalOrdinal, Node>::local_map_type LocalMap;
+  GlobalOrdinal nentries = gids.extent(0);
+  typedef Kokkos::Bitset<device_type> bitset_t;
+  typedef Kokkos::ConstBitset<device_type> const_bitset_t;
+  //Determine the set of GIDs in the column map using a dense bitset (covers entire possible range)
+  bitset_t presentGIDs(maxCol - minCol + 1);
+  Kokkos::parallel_for(range_type(0, nentries), ColMapFunctors::GatherPresentEntries<GlobalOrdinal, GView, bitset_t>(minCol, gids, presentGIDs));
+  execution_space().fence();
+  const_bitset_t constPresentGIDs(presentGIDs);
+  GView mapGIDs("Sum colmap GIDs", constPresentGIDs.count());
+  std::cout << "Rank " << myRank << ": determined there are " << mapGIDs.extent(0) << " unique global cols on this proc.\n";
+  Kokkos::parallel_scan(range_type(0, maxCol - minCol + 1), ColMapFunctors::ListGIDs<GlobalOrdinal, GView, const_bitset_t>(minCol, mapGIDs, constPresentGIDs));
+  execution_space().fence();
```

Comment:
You probably need the fence here because Map's constructor needs to access the data on host, but perhaps we should move the fence into Map's constructor.